### PR TITLE
Update meeting

### LINKED ISSUE #94: Update meeting
- Provide the creator to update a meeting
    - Add/ remove participants 
    - Add/remove optional meeting dates (before the meeting has been set)
    - Change the description, color, location
- Adding tests

### PR DESCRIPTION
**Pull request by YaelD**
- Adding some code to handle a get request of the meeting creator in the views file.
- Adding some code to handle a post request of the meeting creator in the views file.
- Adding a URL for updating a meeting.
- Adding a template file that contains the meeting details and will be shown to the meeting's participants
- Adding code handling with adding and removing participants from a meeting
- Adding code handling with rendering different update-meeting templates to the creator and participants
- Adding code handle with meeting DateTime changes and updating the algorithm timeout
- Updating algorithm reminders while the timeout changes

- fix: #94

### CODE REVIEW

**Comment by kobihk** [comment]
@YaelD please fix the conflicts

File: events/forms.py [hunk 1]
```diff
@@ -90,7 +136,13 @@ def clean(self):
                 if (start_date, end_date) in dates:
                     raise ValidationError("The optional meeting dates should be different")
                 if start_date < min_time_to_set_meeting:  # check min time of 1 hour to a new meeting date
+                    print(f"{start_date} < {min_time_to_set_meeting}")
                     raise ValidationError("Meeting can be set only one hour later from now")
+                else:
+                    if self.creation_meeting_time:
+                        if start_date < timezone.now():
+                            print("in past now")
```
Comment (MortarDefender) [inline events/forms.py]: Can you remove the print function 😄

File: events/forms.py [hunk 2]
```diff
@@ -77,9 +115,17 @@ def clean(self):
 
         dates = []
         min_time_to_set_meeting = timezone.now() + timedelta(hours=1)
+
+        if self.creation_meeting_time:
+            min_time_to_set_meeting = self.creation_meeting_time + timedelta(hours=1)
+
+            print(self.creation_meeting_time)
+            print(min_time_to_set_meeting)
```
Comment (MortarDefender) [inline events/forms.py]: Can you remove the print function 😄

File: events/planner.py [hunk 3]
```diff
@@ -187,14 +198,26 @@ def send_invite_notification(meeting_id):
 
     @staticmethod
     def send_meeting_invitation_email_to_participants(meeting_id, meeting_creator):
+        email_title, message = EventPlanner.create_invitation_to_meeting_message(meeting_id, meeting_creator)
+        all_meeting_participants = EventParticipant.objects.get_an_event_participants_without_creator(meeting_id)
+        print(all_meeting_participants)
```
Comment (MortarDefender) [inline events/planner.py]: Can you remove the print function 😄

File: events/static/events/create_event.css [hunk 4]
```diff
@@ -229,4 +292,4 @@ a {
 
 .fa-file:hover {
     color: #aae0f0;
-}
+}
```
Comment (MortarDefender) [inline events/static/events/create_event.css]: Can you add a new line here 😄

File: users/static/site_notification/notification.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ function build_table() {
 	$.get(get_notification, {}, function(data, status) {
 		
 		$(id).empty();
-
+		console.log(data);
```
Comment (MortarDefender) [inline users/static/site_notification/notification.js]: Can you remove the print function 😄

**Review by MortarDefender — CHANGES_REQUESTED** [review]

File: events/forms.py (hunk 1 above)
Comment (YaelD) [reply events/forms.py]: Done

File: events/forms.py (hunk 2 above)
Comment (YaelD) [reply events/forms.py]: Done 👍

File: events/planner.py (hunk 3 above)
Comment (YaelD) [reply events/planner.py]: Done 👍

File: events/static/events/create_event.css (hunk 4 above)
Comment (YaelD) [reply events/static/events/create_event.css]: done 👍

File: users/static/site_notification/notification.js (hunk 5 above)
Comment (YaelD) [reply users/static/site_notification/notification.js]: Done 👍

File: users/templates/user_site/home.html [hunk 6]
```diff
@@ -154,6 +156,46 @@ <h1 style="font-family: 'Bodoni Moda', serif; font-size: 4.8em;">{{ month_name }
                                 {{ day.2 }}
                             </p>
 
+<<<<<<< HEAD
```
Comment (RonHachmon) [inline users/templates/user_site/home.html]: Please delete git comments

File: events/views.py [hunk 7]
```diff
@@ -117,74 +129,160 @@ def __init__(self, **kwargs) -> None:
...
-                                        int(year), int(month), int(day), current_time.hour, current_time.minute
-                                    ).strftime("%Y-%m-%dT%H:%M")
-            }
+            self.formset_meeting_data = list(
+                OptionalMeetingDates.objects.get_all_event_dates(meeting_id).exclude(
+                    date_time_start=meeting_instance.date_time_start, date_time_end=meeting_instance.date_time_end
+                    )
+                )
+
+            for index, possible_date in enumerate(self.formset_meeting_data):
```
Comment (RonHachmon) [inline events/views.py]: can you please create a different method for lines 193 -198

File: events/views.py [hunk 8]
```diff
@@ -117,74 +129,160 @@ def __init__(self, **kwargs) -> None:
...
+    def calc_timeout_and_send_meeting_notifications(request, event_instance, event_creator,
+                                                    updated_meeting, min_date_change):
+        timeout = EventPlanner.get_timeout(event_instance)
+
+        if updated_meeting and min_date_change:
+            Reminder.objects.remove_algorithm_reminder(event_creator)
+            EventPlanner.creating_meeting_reminders(event_creator, timeout)
+
+        meeting_status = "created"
+        if updated_meeting:
```
Comment (RonHachmon) [inline events/views.py:291]: please move it to one line 
```suggestion
         meeting_status = "created" if updated_meeting else "updated"
```

File: events/views.py [hunk 9]
```diff
@@ -229,18 +339,59 @@ def saving_all_optional_meeting_dates(self, event_creator, event_instance, optio
...
-            event_instance.delete()
+            if not meeting_instance:
+                event_instance.delete()
             return False
         return True
 
     @staticmethod
-    def check_participant_formset(request, event_instance, meeting_participants_formset):
+    def check_participant_formset(request, event_instance, meeting_participants_formset, meeting_instance=None):
+        event_participants = EventParticipant.objects.filter(event_id=event_instance, is_creator=False)
```
Comment (RonHachmon) [inline events/views.py]: please use 
https://github.com/redhat-beyond/lets-meet/blob/bcf5116b942c7cdd5b33752eb6d024b0ffeae5dc/events/class_models/participant_model.py#L12-L13

File: events/views.py [hunk 10]
```diff
@@ -229,18 +339,59 @@ def saving_all_optional_meeting_dates(self, event_creator, event_instance, optio
...
+        old_participants = [(participant.event_id, participant.user_id) for participant in event_participants]
+        meeting_creator = EventParticipant.objects.get(event_id=event_instance, is_creator=True)
```
Comment (RonHachmon) [inline events/views.py]: please use 
https://github.com/redhat-beyond/lets-meet/blob/bcf5116b942c7cdd5b33752eb6d024b0ffeae5dc/events/class_models/participant_model.py#L15-L16

File: events/views.py [hunk 11]
```diff
@@ -251,17 +402,147 @@ def check_participant_formset(request, event_instance, meeting_participants_form
...
+        return datetime.fromtimestamp(
+            timeout.timestamp() - difference
+            ).replace(tzinfo=timezone.get_current_timezone()).replace(second=0).replace(microsecond=0)
+
+
+@login_required(login_url=LOGIN_PAGE)
+def show_meeting(request, meeting_id):
+
+    try:
+        participant = EventParticipant.objects.get(user_id=request.user, event_id=meeting_id)
```
Comment (RonHachmon) [inline events/views.py]: please use 
https://github.com/redhat-beyond/lets-meet/blob/bcf5116b942c7cdd5b33752eb6d024b0ffeae5dc/events/class_models/participant_model.py#L18-L19

File: events/views.py [hunk 12]
```diff
@@ -251,17 +402,147 @@ def check_participant_formset(request, event_instance, meeting_participants_form
...
+
+    MeetingParticipantsFormset = formset_factory(
+        ParticipantForm, formset=BaseParticipantFormSet,
+        max_num=10, extra=0
+    )
+
+    if request.method == 'POST':
+        try:
+            event_instance = Event.objects.get(id=meeting_id)
+            is_creator = EventParticipant.objects.get(user_id=request.user, event_id=event_instance).is_creator
```
Comment (RonHachmon) [inline events/views.py]: please use
 https://github.com/redhat-beyond/lets-meet/blob/bcf5116b942c7cdd5b33752eb6d024b0ffeae5dc/events/class_models/participant_model.py#L18-L19

File: events/views.py [hunk 13]
```diff
@@ -251,17 +402,147 @@ def check_participant_formset(request, event_instance, meeting_participants_form
...
+
+    return redirect('show_meeting', meeting_id=meeting_id)
+
+
+@login_required(login_url=LOGIN_PAGE)
+def get_meeting_participants(request, meeting_id):
+    user = request.user
+
+    try:
+        participant = EventParticipant.objects.get(user_id=user, event_id=meeting_id)
```
Comment (RonHachmon) [inline events/views.py]: please use
https://github.com/redhat-beyond/lets-meet/blob/bcf5116b942c7cdd5b33752eb6d024b0ffeae5dc/events/class_models/participant_model.py#L18-L19

**Review by RonHachmon — CHANGES_REQUESTED** [review]

**Comment by kobihk** [comment]
@YaelD please fix the conflicts and push the CR to make sure we can merge it before the demo ;-)

**Comment by YaelD** [comment]
> @YaelD please fix the conflicts and push the CR to make sure we can merge it before the demo ;-)

@kobihk  
I'm on it 👍

File: users/templates/user_site/home.html (hunk 6 above)
Comment (YaelD) [reply users/templates/user_site/home.html]: Done 👍

File: events/views.py (hunk 7 above)
Comment (YaelD) [reply events/views.py]: Done 👍

File: events/views.py (hunk 8 above)
Comment (YaelD) [reply events/views.py:291]: There is more logic that depends on the `if updated_meeting` condition besides changing the value of `meeting_status`.

File: events/views.py (hunk 9 above)
Comment (YaelD) [reply events/views.py]: Done, thanks 👍

File: events/views.py (hunk 10 above)
Comment (YaelD) [reply events/views.py]: Done, thanks 👍

File: events/views.py (hunk 11 above)
Comment (YaelD) [reply events/views.py]: Done, thanks 👍

File: events/views.py (hunk 12 above)
Comment (YaelD) [reply events/views.py]: Done, thanks 👍

File: events/views.py (hunk 13 above)
Comment (YaelD) [reply events/views.py]: Done, thanks 👍

**Review by RonHachmon — DISMISSED** [review]
Looks great 😄

**Review by RonHachmon — DISMISSED** [review]

**Review by MortarDefender — APPROVED** [review]
LGTM

**Review by BelikovYoav — APPROVED** [review]

**Review by kobihk — APPROVED** [review]

**Pull request merged**